### ISSUE TITLE
Cluster startup deadlock involving Lucene static initialization bits

### ISSUE THREAD

**Issue by shikhar** [elastic/elasticsearch#11170]
From https://issues.apache.org/jira/browse/LUCENE-6482

> during startup of Elasticsearch nodes: From the last 2 stack traces, you see that there are 2 things happening in parallel: Loading of Codec.class (because an Index was opened), but in parallel, Elasticsearch seems to initialize the PostingsFormats.class in the class CodecModule (Elasticsearch). In my opinion, this should not happen in parallel, but a fix would maybe that CodecModule should also call Codecs.forName() so those classes are initialized sequentially at one single place. The problem with Codec.class and PostingsFormat.class clinit running in parallel in different threads may have similar effects like you have seen in the blog post (Codecs depend on PostingsFormat and some PostingsFormats depend on the Codec class, which then hangs if PostingsFormats and Codecs are initialized from 2 different threads at same time, waiting for each other). But we have no chance to prevent this (unfortunately).

> I cannot say for sure, but something seems to be fishy while initializing Elasticsearch, because there is too much happening at the same time. In my opinion, Codecs and Postingsformats and Docvalues classes should be initialized sequentially, but I have no idea how to enforce this.

**Comment by clintongormley** [comment]
@shikhar We haven't seen this issue reported before.  From your last comment:

> It might have been due to using a custom Elasticsearch discovery plugin which is purely asynchronous that those 2 bits ended up happening in parallel, and caused the deadlock.

Did you determine whether this happened when you weren't using the custom plugin?

**Comment by shikhar** [comment]
@clintongormley Not yet, but I'll try to ascertain whether this is still happening by doing a ton of cluster restarts with and without the plugin. I'll reopen if I find this is still an issue. Thanks :)

**Comment by shikhar** [comment]
I have been able to reproduce this with Zen. The only interesting setting is probably `discovery.zen.publish_timeout=0` which makes it similar to [eskka](http://github.com/shikhar/eskka) in that master does not block for acks before processing more updates. But I'm not sure if this is relevant to the problem at all, just thought I'd mention in case it is.

To describe what happens, the cluster is stuck in the RED state. On the master node, there are many errors like:
```
[2015-05-19 12:48:34,460][WARN ][gateway.local            ] [search45-es1] [mfg-1431690242][0]: failed to list shard stores on node [vrHV_EVZSuSFdCGEC7lUsg]
org.elasticsearch.action.FailedNodeException: Failed node [vrHV_EVZSuSFdCGEC7lUsg]
        at org.elasticsearch.action.support.nodes.TransportNodesOperationAction$AsyncAction.onFailure(TransportNodesOperationAction.java:206)
        at org.elasticsearch.action.support.nodes.TransportNodesOperationAction$AsyncAction.access$1000(TransportNodesOperationAction.java:97)
        at org.elasticsearch.action.support.nodes.TransportNodesOperationAction$AsyncAction$4.handleException(TransportNodesOperationAction.java:178)
        at org.elasticsearch.transport.TransportService$TimeoutHandler.run(TransportService.java:529)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.elasticsearch.transport.ReceiveTimeoutTransportException: [search44-es1][inet[/172.31.240.55:8301]][internal:cluster/nodes/indices/shard/store[n]] request_id [4273] timed out after [30001ms]
        ... 4 more
```

A thread dump from the `vrHV_EVZSuSFdCGEC7lUsg` node that it is complaining about: 
https://gist.github.com/shikhar/8ad5c166c6a20458c4d2 -- grepping for `Codec` or `<init>` should reveal the threads where the static initialization deadlock is happening. These threads are shown as RUNNABLE while in `Object.wait()`

My hunch is that this can happen while the cluster is starting up and an index create event lands (we have some crons that are creating new indexes in the background as a way of reindexing). You can see that one of the threads that is deadlocked landed in the Codec-ey bits via:
```
org.elasticsearch.indices.cluster.IndicesClusterStateService.applyNewIndices(IndicesClusterStateService.java:311)
```

**Comment by shikhar** [comment]
I think the initialization of the deadlock-prone bits that @uschindler mentioned should happen in a deterministic manner at ES startup, rather than e.g. when creating indexes.

**Comment by clintongormley** [comment]
@shikhar thanks for coming back to us with the details

**Comment by shikhar** [comment]
Forgot to mention that this was reproduced on ES 1.5.2

**Comment by uschindler** [comment]
The discussion when and why this might happen is here:
https://issues.apache.org/jira/browse/LUCENE-6482

The proposal is to load the Codec, PostingsFormat and DocValuesFormat in the Elasticsearch Boostrap class. So something like calls to:

```java
Codec.reloadCodecs();
PostingsFormat.reloadPostingsFormats();
// and others, see Solr's SolrResourceLoader startup
```

somewhere at the beginning of Boostrap class. In Solr this is already done in the startup code. This prevents dependency problems if those index classes are started in a concurrent way.

**Comment by rmuir** [comment]
Uwe, why should users of lucene have to do this? Tracking down a class loading deadlock like this is something i do not wish on anyone (thank you very much @shikhar for digging here). So I don't like that you need a "magical sequence" at the startup of your app, or you might get hangs.

Is there no better way?

**Comment by uschindler** [comment]
The problem here is that on Elasticsearch startup the big builder is initializing a lot of shit in a unforseeable order, and concurrent in addition. The main problem here is that the Codec/Postingsformat/... stuff has a large spaghetti of load-time dependencies, so if called in the wrong order concurrently, it hangs.

To solve the issue here, my suggestion is to simply call those static initializes at a defined place in Elasticsearch's bootstrap, so it cannot happen that index A is initialized at the same time while index B is also booting up and both are loading SimpleText codec and ElasticsearchFooBar codec in parallel.

We can then look into Lucene's code to cleanup the loading of codecs. The big problem is the dependency graph and static initializers. The main problem is the Codec.getDefault() value that is initialized with Codec.forName(). I think the problem would be much easier if we would just initialize the default codec with a simple new LuceneXYCodec() instead of SPI on the <clinit> phase. Currently we have code like "static Codec defaultCodec = forName("Lucene50");"

So two steps:
- prevent deadlocks in ES caused by Lucene for now by explicitely initializing Codec class in Bootstrap.java
- change Lucene's Codec/PostingsFormat to initialize the default codecs/formats without forName(), because this can lead to deadlock. Just use a simple "new" call to constructor.

Similar propblems are regularirly happening in ICU4J: You remember the case where ICU4J caused NullPointerExceptions because it was trying to print a message using the default codec before the default codec was actually loaded and initialized... The problems in Lucene are the same, just happening in multithreaded code.

**Comment by rmuir** [comment]
> I think the problem would be much easier if we would just initialize the default codec with a simple new LuceneXYCodec() instead of SPI on the phase. Currently we have code like "static Codec defaultCodec = forName("Lucene50");"

Is that really all thats needed to fix it in Lucene? If so, +1.

But for master branch, we shouldn't need a hack in bootstrap, lets just fix it in lucene and upgrade to a newer lucene snapshot jar.

For any backport fixes (elasticsearch 1.X) your solution is practical. But i would not be upset to see a 4.10.5 either.

**Comment by uschindler** [comment]
In my opinion this should work fine. We should just not use SPI to initialize the defaults.

The problem with forName() is: You are loading a lot of classes by that, which initialize themselves and may reference to the class doing the forName(). If we initialize the default codec statically, we dont need to wait for forName() to complete, the default codec is there ASAP. If some index file needs another codec, it can load it with forName() but there is no risk in deadlocking.

**Comment by rmuir** [comment]
I see. I think this is a lot simpler in 5.x, because forName is not "abused" to load "impersonator" codecs in tests. You remember, in that case we had a Lucene3x that was writeable and we reordered jar files in tests and all that... all gone now.

So I think its too tricky to fix as a bugfix in lucene 4.10.x, and we should apply your logic to bootstrap for ES 1.x. But lets fix lucene 5.x as proposed, then simply update our snapshot for ES master.

**Comment by uschindler** [comment]
I would wish we could write a test to check this out... But as usual, this test would need to run isolated (new and fresh JVM trying to open several indexes in parallel without loading Codec class before).

Thank you for remembering me why we did this Codec.forName() for the static initialization! That makes sense. I think we should kill this in 5.x, its a one-line change (and maybe a change in smoketester that validates that the defaultCodec line was updated after release). PostingsFormat is not affected, all of that is caused by Codec's <clinit>.

I analyzed the stack traces provided. What happens in the example is exactly as described: One of the threads is initializing Codec class (<clinit>), as side effect of opening an index. 3 other threads are opening other indexes at same time. Because the Codec class is not yet initialized in clinit, those threads are waiting before SegmentInfos's call to Codec.forName(), which is blocked because the SegmentInfos's call cannot access the Codec class yet. Because of concurrent class loading the newInstance() call in NamedSPILoader to the codec class actually loaded is blocked (too). This happens because JVM has internal locks on that (not 100% sure why and where this blocks, but seems to be the reason).

**Comment by rmuir** [comment]
Related to that, there was some previous discussion about implementing a static detector here: https://issues.apache.org/jira/browse/LUCENE-5573

I am just hoping some policeman finds himself bored one day and implements the ASM logic so we could try to detect these, even if it wouldn't find this particular one.

**Comment by uschindler** [comment]
I know this issue :-) I wonder that there is no Eclipse plugin that can detect this!

**Comment by rmuir** [comment]
I did a quick survey, nothing exhaustive but I couldn't find anything. I haven't tried to play around with writing a detector. But IMO ideally it would be a policeman-tool like forbidden API, and we just scan for it in builds and fail the build.

**Comment by shikhar** [comment]
hey @rmuir what are the plans for tackling this? It is fixed in Lucene 5.2.1 (https://issues.apache.org/jira/browse/LUCENE-6482) -- if only ES 2.0 can use 5.x, there probably either needs to be that ES Bootstrap workaround or a backport to Lucene 4.x

**Comment by uschindler** [comment]
I would go with the Bootstrap workaround for ES 1.5 and 1.6. This is very easy to implement and there is no need to release several 4.x versions of Lucene. Should I provide a PR?

**Comment by clintongormley** [comment]
@uschindler yes please! :)

**Comment by s1monw** [comment]
@uschindler can you open a PR for this against 1.x branch please

**Comment by uschindler** [comment]
OK, will provide one later! I am currently in the U.S. and busy :-)

**Comment by uschindler** [comment]
Ah which branch?

**Comment by uschindler** [comment]
Found it out: 1.x

**Comment by uschindler** [comment]
I created a PR: #11837

This one does a "fake" Codecs.availableCodecs() on InternalNode startup. I did not do it in the Boostrap, so also people embedding ES can make use of it. In any case, the class is cheap and does not slowdown, it just returns an immutable list.

Is there a problem that InternalNode directly depends on lucene-core.jar?

**Comment by uschindler** [comment]
Thanks for merging!
